### PR TITLE
feat(gdn): add fused chunk-parallel prefill path

### PR DESCRIPTION
**Pull request by Infra-lab-dev**
## Motivation

Qwen3.5 Gated DeltaNet (GDN) carries recurrent state across prompt tokens. In the original token-by-token JAX prefill path, that recurrence serialized work and repeatedly materialized intermediate state, making long-context prefill a major latency and TTFT bottleneck.

This PR addresses that bottleneck with a TPU/Pallas kernel that fuses causal Conv1D with chunk-parallel GDN prefill. The fused path reduces kernel launches and intermediate state traffic while preserving packed-request handling, recurrent-state-pool semantics, and the existing base-JAX decode path.

After #1505, the native prefill implementation is `chunked_jax`. This PR keeps that route available and makes `fused_chunk_parallel` the default. The profiling results below compare against the pre-#1505 `token_scan` path and are retained as historical evidence; they are not a measured comparison with `chunked_jax`.

## Changes

- Add the Pallas fused Conv1D+GDN prefill kernel from TPU-Inference v3.
- Add `--gdn-prefill-impl` to select `chunked_jax` or `fused_chunk_parallel`; the fused route is the default, and both routes keep base-JAX decode.
- Integrate the fused path with packed requests and recurrent state pools, with startup capability checks and correctness tests.

## Accuracy Tests

- The serving A/B matched all 64 output tokens across ten measured rounds, with
  no request errors or non-finite results.
- TPU tests cover lengths `1, 63, 64, 65, 127, 128, 129`, state updates,
  packed and zero-length requests, and checkpoint isolation. The local batches
  passed with `53`, `10`, and `21` tests; the 16-host contract run also
  passed.
- All six historical `token_scan`/`fused_chunk_parallel` profiling cells passed
  correctness, finite-value, shape, input-hash, route, and source-HEAD checks.

## Benchmarking and Profiling

Timing and profiling were collected separately. XProf trace durations are used
only for internal operator attribution and do not replace the formal timing
table.

The `token_scan` results below were captured before #1505 landed. They remain
historical evidence for the recorded source revisions; this PR does not claim a
measured comparison against the current `chunked_jax` native path.

### Qwen3.5-397B-A17B full-serving A/B

This is a separate full-serving validation, not the single-device kernel
profiling target.

Configuration:

- Model revision: `8472618112abcbd45acbcdc58436aff4233c23f7`
- BF16, TPU v7x64 (`4x4x4`, 16 hosts)
- JAX/jaxlib `0.10.2`, libtpu `0.0.43`
- TP/DP/EP `128/8/128`
- Five measured rounds per cell, 80 requests per round
- Exact 64-token output identity between variants

| Workload | Metric | `token_scan` | `fused_chunk_parallel` | Change |
|---|---|---:|---:|---:|
| 4096 to 1, concurrency 16 | Input throughput | 9,485.999241 tok/s | 40,821.238522 tok/s | +330.331455% |
| 4096 to 1, concurrency 16 | Mean TTFT | 6,515.492437 ms | 1,513.621697 ms | -76.768883% |
| 1024 to 100, concurrency 16 | Output throughput | 275.949103 tok/s | 398.225380 tok/s | +44.311170% |
| 1024 to 100, concurrency 16 | Mean TPOT | 36.643958 ms | 36.005768 ms | -1.741596% |

The decode implementation is unchanged. The decode-cell improvement includes
the reduced prefill/TTFT portion of those requests.

### Qwen3.5-397B-A17B GDN kernel timing

This is a single-device kernel benchmark, separate from the full-serving A/B.

- Revision: `31b8ef0d3348f9260a37140949802e3d7efe94cb`
- Environment: one v7x JAX device, BF16, JAX/jaxlib `0.10.2`, libtpu
  `0.0.42.1`; TP16 local shape
  `n_kq=1, n_v=4, d_k=d_v=128, mixed_qkv=768, conv_kernel=4`
- Prefill: 5 warmups followed by 30 measured runs
- Decode: 16K cached state, BS64, 10 warmups followed by 64 one-token runs

Both routes use the same deterministic inputs and initial state, with separate
processes and compilation caches. Compilation and warmup are excluded, results
are synchronized with `block_until_ready()`, and all raw samples are kept.

| Workload | `token_scan` median [P10, P90] | `fused_chunk_parallel` median [P10, P90] | fused vs token scan |
|---|---:|---:|---:|
| Prefill 4K | 30.987 [30.494, 31.191] ms | 1.352 [1.262, 1.450] ms | 22.91x |
| Prefill 32K | 411.064 [410.855, 411.323] ms | 12.679 [12.550, 12.901] ms | 32.42x |
| Decode 16K cache, BS64 | 1.236 [1.002, 1.357] ms | 1.269 [1.229, 1.370] ms | 0.974x (-2.66%) |

The formal table contains 248 retained `token_scan`/`fused_chunk_parallel`
samples.

### XProf and compiler attribution

Profiling produced non-empty XPlanes and source mapping for all six retained
historical `token_scan`/`fused_chunk_parallel` cells. HLO/compiler dumps were retained for every cell, with
applicable Pallas LLO retained where the route produced it.

| Cell | HLO self / call | Dominant categories | Top XProf evidence |
|---|---:|---|---|
| Prefill 4K `token_scan` | 21.059 ms | elementwise 89.5%; matmul 9.4% | `select_select_fusion.4` 180.089 ms total; two dynamic slices about 79 ms each |
| Prefill 4K `fused_chunk_parallel` | 0.203 ms | custom kernel 80.8%; formatting 13.9% | `fused_conv1d_gdn_per_seq.1` 4.908 ms |
| Prefill 32K `token_scan` | 73.825 ms | elementwise 94.7%; matmul 4.6% | select/dynamic-slice operations occupy the top five hotspots |
| Prefill 32K `fused_chunk_parallel` | 10.278 ms | custom kernel 98.0%; formatting 1.5% | `fused_conv1d_gdn_per_seq.1` 302.268 ms |
| Decode `token_scan` | 0.0783 ms | state custom fusion 46.3%; formatting 22.6% | scatter/copy/gather and recurrence fusions dominate |
| Decode `fused_chunk_parallel` | 0.0899 ms | custom kernel 38.4%; formatting 32.0%; elementwise 19.7% | fused batched call 2.138 ms; copy groups 1.385 ms total |

Notes from XProf:

- `token_scan` prefill spends most of its time in elementwise work and state
  materialization. In `fused_chunk_parallel`, the fused kernel accounts for
  about 98% of 32K HLO self-time, matching the large timing improvement.
- The `fused_chunk_parallel` decode result is kernel-only; production decode
  still uses the existing reference decode path. At one token per sequence, state traffic and copy/formatting
  overhead are not amortized.

## Community Tracking

Addresses #1416 and contributes to the Qwen3.5 performance roadmap tracked by
#1421. It builds on the existing GDN kernel/backend foundation from #1059.

Issue #1416 originally proposed chunked JAX followed by an in-tree KDA/Pallas
path. After profiling and comparing the token-scan baseline, Chunkwise KDA,
and fused chunk-parallel implementations, this PR selects the fused
chunk-parallel Conv1D+GDN prefill path as the production candidate.

PR #1466 was the earlier Chunkwise KDA implementation candidate. This PR
supersedes #1466 for the prefill-performance work tracked by #1416.

## Tracking

Tracks https://github.com/primatrix/projects/issues/324.

Current design RFC: https://github.com/primatrix/wiki/pull/204.

Historical Chunkwise design: https://github.com/primatrix/wiki/pull/201.

PR #1466 was the earlier implementation candidate for #324. It replaced the
token-by-token GDN prefill recurrence with the existing SGL-JAX `chunk_kda`
path and established the initial correctness and performance baseline.

Subsequent route comparison and Qwen3.5-397B-A17B validation selected the
fused chunk-parallel Conv1D+GDN prefill path as the production candidate.
This PR therefore supersedes #1466 as the proposed implementation for #324.

This PR is not stacked on or dependent on #1466 and does not include the
Chunkwise KDA adapter from #1466.

## Checklist

- [x] The PR body is written in English.
- [x] The purpose and tracking issue are documented.
- [x] Capability, provenance, verification, and benchmark documentation is
  included.

### CODE REVIEW

**Review by cjx0709 — COMMENTED** [review]
shoulde not use folder name or arg name like `tpu_inference` or `reference`, it doesn't mean anything, will not help any contributor to understand

File: docs/developer_guide/gdn_tpu_inference_v3.md [hunk 1]
```diff
@@ -0,0 +1,172 @@
+# TPU-Inference v3 GDN Prefill
```
Comment (cjx0709) [inline docs/developer_guide/gdn_tpu_inference_v3.md]: It is not appropriate for this file to appear in the development guide.

File: docs/features/server_arguments.md [hunk 2]
```diff
@@ -178,6 +178,36 @@ For deeper scheduler behavior, see [Scheduler](../architecture/03-scheduler.md)
...
+```bash
+SGLANG_JAX_GDN_PREFILL_IMPL=tpu_inference_v3 \
+python3 -u -m sgl_jax.launch_server \
+  --model-path <QWEN3_5_MODEL_PATH> \
+  --device tpu \
+  --dtype bfloat16 \
+  <OTHER_SERVE_ARGUMENTS>
+```
+
+See [TPU-Inference v3 GDN Prefill](../developer_guide/gdn_tpu_inference_v3.md)
```
Comment (cjx0709) [inline docs/features/server_arguments.md]: also see as the comment before

File: python/sgl_jax/srt/kernels/gdn/tpu_inference_adapter.py [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+    state_indices = metadata.recurrent_indices
+    has_initial_state = metadata.has_initial_state
+    track_indices = metadata.recurrent_track_indices
+    track_mask = metadata.recurrent_track_mask
+    if (track_indices is None) != (track_mask is None):
+        raise ValueError("track_indices and track_mask must either both be set or both be None.")
+    dp = int(backend.mesh.shape.get("data", 1))
+    if conv_state.shape[0] % dp:
+        raise ValueError(f"state pool size {conv_state.shape[0]} must be divisible by DP={dp}.")
+    invalid_track = _validate_track_indices_per_dp(
```
Comment (cjx0709) [inline python/sgl_jax/srt/kernels/gdn/tpu_inference_adapter.py]: Could we move this validation to the metadata producer, or make it debug-only? Under JIT it builds pairwise B×B checks here, and `fused_conv1d_gdn_prefill()` repeats the same validation inside every shard, adding runtime work to the prefill path this PR is optimizing.

File: python/sgl_jax/srt/layers/attention/linear/gdn_backend.py [hunk 4]
```diff
@@ -114,6 +124,38 @@ def __init__(
                 f"of num_k_heads={num_k_heads} (GQA repeat factor)."
             )
 
+        self.requested_prefill_impl = os.environ.get("SGLANG_JAX_GDN_PREFILL_IMPL", "reference")
+        if self.requested_prefill_impl not in {"reference", "tpu_inference_v3"}:
+            raise ValueError(
+                "SGLANG_JAX_GDN_PREFILL_IMPL must be one of 'reference' or 'tpu_inference_v3'."
+            )
+        self.fallback_reason = None
```
Comment (cjx0709) [inline python/sgl_jax/srt/layers/attention/linear/gdn_backend.py]: Could we collapse this to a single selected implementation? Unsupported configurations raise instead of falling back, so `requested` and `effective` are always identical and `fallback_reason` is always `None`; the extra state and logging represent a path that cannot occur.

File: python/sgl_jax/test/test_gdn_tpu_inference_prefill_dp.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+import pytest
+from jax.experimental import multihost_utils
+
+pytestmark = pytest.mark.skipif(
+    not any(device.platform == "tpu" for device in jax.local_devices()),
+    reason="the distributed GDN contract requires real TPU hardware",
+)
+
+
+def test_all_hosts_form_unique_process_and_device_partitions():
```
Comment (cjx0709) [inline python/sgl_jax/test/test_gdn_tpu_inference_prefill_dp.py]: Could we remove this test or have it execute the GDN adapter under DP? It only checks JAX process/device accounting, so it cannot catch a DP regression in this feature.

File: python/sgl_jax/test/test_gdn_tpu_inference_prefill_dispatch.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    with pytest.raises(GDNPrefillCapabilityError, match="TPU"):
+        _backend(monkeypatch, impl="tpu_inference_v3", dtype=jnp.bfloat16)
+
+
+def test_production_constructor_passes_model_dtype_to_backend():
+    source = (
+        Path(__file__).parents[1] / "srt" / "layers" / "attention" / "hybrid_linear_attn_backend.py"
+    ).read_text()
+
+    assert "dtype=runner.model_config.dtype" in source
```
Comment (cjx0709) [inline python/sgl_jax/test/test_gdn_tpu_inference_prefill_dispatch.py]: Could we test the constructor behavior instead of searching the source text? This can pass if the string moves into a comment or dead branch and fail after a harmless refactor, so it does not verify that the production path actually forwards the dtype.

File: test/srt/run_suite.py [hunk 7]
```diff
@@ -384,15 +415,33 @@ def run_one_file(
...
+            "python/sgl_jax/test/layers/test_lightning_backend_dp.py",
+            1,
+            runner="pytest",
+        ),
         TestFile("python/sgl_jax/test/test_kda_attention.py", 6.5),
         TestFile("python/sgl_jax/test/test_kda_attention_dp.py", 6),
         TestFile("python/sgl_jax/test/test_gdn_attention.py", 6.5),
         TestFile("python/sgl_jax/test/test_gdn_attention_dp.py", 6),
+        TestFile(
+            "python/sgl_jax/test/test_gdn_tpu_inference_prefill.py",
```
Comment (cjx0709) [inline test/srt/run_suite.py]: Could we move this numerical test to `unit-test-tpu-v6e-1`? It never creates a multi-device mesh and only invokes the single-device fused kernel, while v6e-4 should be reserved for coverage that actually exercises DP/TP sharding.

File: docs/developer_guide/gdn_tpu_inference_v3.md (hunk 1 above)
Comment (Infra-lab-dev) [reply docs/developer_guide/gdn_tpu_inference_v3.md]: Removed this file in d232a216.

File: python/sgl_jax/test/test_gdn_tpu_inference_prefill_dispatch.py (hunk 6 above)
Comment (Infra-lab-dev) [reply python/sgl_jax/test/test_gdn_tpu_inference_prefill_dispatch.py]: Replaced this with a constructor behavior test in d232a216.

**Review by cjx0709 — APPROVED** [review]
